libs/wire-api/src/Wire/API/Password.hs

```diff
@@ -56,14 +103,89 @@ genPassword =
+mkSafePassword = fmap Password . hashPassword . Text.encodeUtf8 . fromPlainTextPassword
 
 -- | Verify a plaintext password from user input against a stretched
 -- password from persistent storage.
 verifyPassword :: PlainTextPassword' t -> Password -> Bool
 verifyPassword plain opaque =
-  let actual = Pass . Text.encodeUtf8 $ fromPlainTextPassword plain
+  let actual = fromPlainTextPassword plain
       expected = fromPassword opaque
-   in verifyPass' actual expected
+   in checkPassword actual expected
+
+hashPassword :: MonadIO m => ByteString -> m Text
+hashPassword password = do
+  salt <- newSalt $ fromIntegral defaultParams.saltLength
+  let key = hashPasswordWithSalt password salt
+  pure $
+    Text.intercalate
+      "|"
+      [ "14",
+        "8",
+        "1",
+        Text.decodeUtf8 . B64.encode $ salt,
+        Text.decodeUtf8 . B64.encode $ key
+      ]
+
+hashPasswordWithSalt :: ByteString -> ByteString -> ByteString
+hashPasswordWithSalt password salt = hashPasswordWithParams defaultParams password salt
+
+hashPasswordWithParams ::
+  ( ByteArrayAccess password,
+    ByteArrayAccess salt
+  ) =>
+  ScryptParameters ->
+  password ->
+  salt ->
+  ByteString
+hashPasswordWithParams parameters password salt = convert (generate (fromScrypt parameters) password salt :: Bytes)
+
+checkPassword :: Text -> Text -> Bool
+checkPassword actual expected = fromMaybe False $ do
+  (sparams, salt, hashedKey) <- parseScryptPasswordHashParams $ Text.encodeUtf8 expected
+  let producedKey = hashPasswordWithParams sparams (Text.encodeUtf8 actual) salt
+  pure $ hashedKey `constEq` producedKey
+
+newSalt :: MonadIO m => Int -> m ByteString
+newSalt i = liftIO $ getRandomBytes i
+{-# INLINE newSalt #-}
+
+parseScryptPasswordHashParams :: ByteString -> Maybe (ScryptParameters, ByteString, ByteString)
+parseScryptPasswordHashParams passwordHash = do
+  let paramList = Text.split (== '|') . Text.decodeUtf8 $ passwordHash
+  guard $ length paramList == 5
+  let [ scryptRoundsT,
+        scryptBlockSizeT,
+        scryptParallelismT,
+        salt64,
+        hashedKey64
+        ] = paramList
+  rounds <- readT scryptRoundsT
+  blockSize <- readT scryptBlockSizeT
+  parallelism <- readT scryptParallelismT
+  salt <- from64 salt64
+  hashedKey <- from64 hashedKey64
+  let outputLength = fromIntegral $ C8.length hashedKey
+      saltLength = fromIntegral $ C8.length salt
+  pure
+    ( ScryptParameters {..},
+      salt,
+      hashedKey
+    )
+
+-- | Same as 'read' but works on 'Text'
+readT :: Read a => Text -> Maybe a
+readT = readMaybe . Text.unpack
+{-# INLINE readT #-}
+
+-- | Same as 'show' but works on 'Text'
+showT :: Show a => a -> Text
+showT = Text.pack . show
+{-# INLINE showT #-}
+
+-- | Decodes a base64 'Text' to a regular 'ByteString' (if possible)
+from64 :: Text -> Maybe ByteString
+from64 = toMaybe . B64.decode . Text.encodeUtf8
+  where
+    toMaybe = either (const Nothing) Just
```

Comment:
```suggestion
from64 = hush . B64.decode . Text.encodeUtf8
```

Reply:
I'm not sure it's worth adding the `errors` package as a dependency for this. 🤔

Reply:
there's also one that does the same thing for `MonadError`, I'd guess we already depend on that.

Reply:
https://hackage.haskell.org/package/either-5.0.2/docs/Data-Either-Combinators.html#v:rightToMaybe